File: Find Closest Node to Given Two Nodes.cpp
class Solution {
public:
vector<int> Bfs(int src, vector<int> &edges) {
    queue<int> q;
    int n = edges.size();
    q.push(src);
    vector<int> dist(n, -1);
    dist[src] = 0;
    while (!q.empty()) {
        int cur = q.front();
        q.pop();
        int next = edges[cur];
        if (next != -1) {
            if (dist[next] == -1) {
                dist[next] = dist[cur] + 1;
                q.push(next);
            }
        }
    }
    return dist;
}

int closestMeetingNode(vector<int> &edges, int node1, int node2) {
    vector<int> v1 = Bfs(node1, edges);
    vector<int> v2 = Bfs(node2, edges);
    int n=edges.size();
    int ans=-1;
    int mn=INT_MAX;
    for (int i = 0; i < n; ++i) {
        if(v1[i]!=-1&&v2[i]!=-1){
            if(max(v1[i],v2[i])<mn){
                ans=i;
                mn=max(v1[i],v2[i]);
            }
        }
    }
    return ans;
}

};

File: Find Closest Node to Given Two Nodes.java
class Solution {
    public int closestMeetingNode(int[] edges, int node1, int node2) {
        int n = edges.length;
        int[] dist1 = new int[n];
        int[] dist2 = new int[n];
        Arrays.fill(dist1, Integer.MAX_VALUE);
        Arrays.fill(dist2, Integer.MAX_VALUE);
        dist1[node1] = 0;
        dist2[node2] = 0;
        
        Boolean[] vis1 = new Boolean[n], vis2 = new Boolean[n];
        Arrays.fill(vis1, Boolean.FALSE);
        Arrays.fill(vis2, Boolean.FALSE);

        dfs(edges, node1, vis1, dist1);
        dfs(edges, node2, vis2, dist2);
        
        int minNode = -1, minDis = Integer.MAX_VALUE;
        for(int i=0;i<n;i++){
            if(minDis>Math.max(dist1[i],dist2[i])){
                minNode = i;
                minDis = Math.max(dist1[i],dist2[i]);
            }
        }
        
        return minNode;
    }
    
    public static void dfs(int[] edges, int node, Boolean[] vis, int[] dist){
        vis[node] = true;
        
        int ni = edges[node];
        if(ni!=-1 && vis[ni]==false){
            dist[ni] = 1 + dist[node];
            dfs(edges, ni, vis, dist);
        }
    }
}
